libsolidity/interface/StandardCompiler.cpp

```diff
@@ -588,7 +590,10 @@ std::variant<OptimiserSettings, Json> parseOptimizerSettings(Json const& _jsonIn
 			return *error;
 		if (auto error = checkOptimizerDetail(details, "simpleCounterForLoopUncheckedIncrement", settings.simpleCounterForLoopUncheckedIncrement))
 			return *error;
-		settings.optimizeStackAllocation = settings.runYulOptimiser;
+
+		if (optimizerEnabled || settings.runYulOptimiser)
+			settings.optimizeStackAllocation = true;
```

Comment:
This is wrong, because it means that `enabled: true` + `yul: false` combination enables the flag (on the CLI the flag stays `false` in that case). It should be:

```suggestion
		if (settings.runYulOptimiser)
			settings.optimizeStackAllocation = true;
```

Reply:
Now that I think of it, this means that both on the command line and in Standard JSON we do disable the flag when the optimizer as a whole is disabled. That's not a part of the bug, but I wonder if I should have changed that behavior in #14149. After all, the way it is now means that the state of the flag depends on whether we run evmasm optimizer, which is weird.

We should probably change that in a separate PR.

Reply:
> This is wrong, because it means that `enabled: true` + `yul: false` combination enables the flag (on the CLI the flag stays `false` in that case). It should be:

However, on the CLI, the `optimizeStackAllocation` setting depends on the value of `optimizer.optimizeYul`:

https://github.com/ethereum/solidity/blob/b6486db197e0b577a07554e7480f5bdb709b9c7c/solc/CommandLineParser.cpp#L272-L276

This value is determined either by using `--optimize` without `--no-optimize-yul` *or* by explicitly setting `--optimize-yul`:
https://github.com/ethereum/solidity/blob/b6486db197e0b577a07554e7480f5bdb709b9c7c/solc/CommandLineParser.cpp#L1242-L1246

In contrast, in the Standard JSON, `settings.runYulOptimiser` depends solely on the `details.yul` option:
https://github.com/ethereum/solidity/blob/b6486db197e0b577a07554e7480f5bdb709b9c7c/libsolidity/interface/StandardCompiler.cpp#L587

And it does not depend on `optimizer.enabled`:
https://github.com/ethereum/solidity/blob/b6486db197e0b577a07554e7480f5bdb709b9c7c/libsolidity/interface/StandardCompiler.cpp#L556

Therefore, on the CLI, the flag does not remain `false`. This mismatch was the root cause of the ICE, where we assumed the flag was `false`, but it was actually `true` in `createMetadata`. This is why I attempted to align the behavior in the Standard JSON with the CLI. However, it’s possible that the CLI behavior itself is incorrect and should be fixed instead.

Since the `optimizer` enables the Yul optimizer by default, I believe the behavior in the PR still correct. In other words: the `optimizeStackAllocation` will only be disabled if the user explicitly use `--no-optimize-yul` or `details.yul: false`. But maybe we should change the CLI instead.

This assumption is based on our docs:
```
        // The new Yul optimizer. Mostly operates on the code of ABI coder v2
        // and inline assembly.
        // It is activated together with the global optimizer setting
        // and can be deactivated here.
        // Before Solidity 0.6.0 it had to be activated through this switch.
        "yul": false,
        // Tuning options for the Yul optimizer.
        "yulDetails": {
          // Improve allocation of stack slots for variables, can free up stack slots early.
          // Activated by default if the Yul optimizer is activated.
          "stackAllocation": true,
          ...
        }
```

Reply:
> Since the `optimizer` enables the Yul optimizer by default, the `optimizeStackAllocation` should only be disabled if the user explicitly use `--no-optimize-yul` or `details.yul: false`, right?

No. I am assuming it should be tied to the internal state of the Yul optimizer. I.e. if the sequence runs, optimized stack allocation should happen unless the user explicitly disables it with `optimizeStackAllocation: false`. And after #14149 the sequence always runs.

Currently it's also disabled when you disable the optimizer as a whole (https://github.com/ethereum/solidity/pull/15655#discussion_r1890770871), which is inconsistent with that assumption and IMO should also be changed.

At least this is the assumption I've been making in my description in the issue and in what I reviewed here so far. IIRC Daniel said that it's just an internal detail and we should always have it enabled in optimized compilation eventually and deprecate the setting. But I'm actually not entirely sure how indispensable this flag is for unoptimized compilation (or if it matters at all). On one hand we still want to avoid doing anything unnecessary that would be considered an optimization in that mode. On the other hand the reason for always running UnusedPruner in the first place was to help with stack issues and this flag is aimed at that too. I think kept flipping between those two views in the original issue and it's part of the reason why it ended up being this inconsistent. We should discuss it with Daniel and get this straight.

Reply:
> In contrast, in the Standard JSON, `settings.runYulOptimiser` depends solely on the `details.yul` option:

> And it does not depend on `optimizer.enabled`:

It does depend on `optimizer.enabled`, because that setting determines its default value. Note the `settings = OptimiserSettings::standard();` on the line below :)

BTW, this is exactly what I meant when I complained about defaults depending on defaults depending on defaults. Why is this thing this tricky to reason about. It's just a bunch of boolean flags :)